04-Protocol-Security.md

```diff
@@ -0,0 +1,391 @@
+
+Should the decryption (i.e. authentication code validation) fail at any point, the session must be terminated. 
+
+### 4.5.1 Handshake Act 1: Noise protocol proposal
+Initiator proposes list of concrete noise-protocol dialects
+
+```
+Protocol proposal message
++---------------+------------------------------------------------------------------------------------------------------+
+| u32           |  MAGIC_NUMBER: Currently set to 861033555, whose LE binary representation is b"STR3"                 |
++---------------+------------------------------------------------------------------------------------------------------+
+| SEQ0_32[u32]  |  List of algorithms that the noise protocol should use for the encrypted session                     |
++---------------+------------------------------------------------------------------------------------------------------+
+
+Message length: 5 + *n* * 4 Bytes, where n is the length byte of the SEQ0_32 field, at most 133
+```
+
+
+```
+Protocol variants:
++--------------------------+-------------------------------------------------------------------------------------------+
+| u32 code (LE repr)       |   Official protocol name                                                                  |
++--------------------------+-------------------------------------------------------------------------------------------+
+| 0x53484353 (b"SCHS")     |  Noise_NX_secp256k1_ChaChaPoly_SHA256 (mandatory)                                         |
++--------------------------+-------------------------------------------------------------------------------------------+
+| 0x53474153 (b"SAGS")     |  Noise_NX_secp256k1_AESGCM_SHA256 (optional)                                              |
++--------------------------+-------------------------------------------------------------------------------------------+
+```
+
+Protocol `Noise_NX_secp256k1_ChaChaPoly_SHA256` must be always supported. It's the only variant that can be implemented using primitives from Bitcoin Core.
+
+### 4.5.2 Handshake Act 2: Noise protocol choice
+Responder confirms its choice of algorithm with a simple response:
+
+```
+Protocol choice message
++---------------+------------------------------------------------------------------------------------------------------+
+| u32           |  Server's choice of algorithm                                                                        |
++---------------+------------------------------------------------------------------------------------------------------+
+
+Message length: 4 Bytes
+```
+
+### 4.5.3 Handshake Act 3: NX-handshake part 1 `-> e`
+
+Prior to starting first round of NX-handshake, both initiator and responder initializes handshake variables `h` (hash output), `ck` (chaining key) and `k` (encryption key):
+
+1.  **prolog** is constructed as a serialized sequence of noise-protocol negotiation message with following structure:
```

Comment:
prolog -> prologue

Reply:
Do we want to state that the prologue is "useful when the parties engaged in negotiation prior to the handshake and want to ensure they share identical views of that negotiation", or is that unnecessary since it is stated in the Noise protocol spec?

Reply:
Maybe it's worth mentioning. I'll include that